h2/src/main/org/h2/engine/InternalAuthenticator.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: Alessandro Ventura
+ */
+package org.h2.engine;
+
+import org.h2.api.Authenticator;
+import org.h2.security.auth.AuthConfigException;
+import org.h2.security.auth.AuthenticationException;
+import org.h2.security.auth.AuthenticationInfo;
+
+/**
+ *    Default authentication implementation. It validate user and password internally on database
+ */
+public class InternalAuthenticator implements Authenticator {
```

Comment:
this is a nice design, but not strictly necessary. If it helps to reduce dependencies from the api package to internal classes, I am fine with the "normal" authentication path remaining in classes like Engine, and only calling the custom authentication code if such is configured.

Reply:
InternalAuthenticator has been created to reuse the internal password check without making public validateUserPasswordHash. It can be removed but without it external authenticators cannot failback explicitly to internal authentication system

Reply:
(a) One possibility is external authenticators returning a result code that tells H2 to use normal authentication. 
(b) Otherwise we will just have to accept that external authenticators will rely on H2's internals, and will therefore not be as stable, i.e. the same situations as custom table engines.

Since it looks like the external auth wants to talk to various bits of H2, it looks like we will to go with (b).

Reply:
I will remove internal authenticator, moving back the login inside Engine.openSession when AUTHREALM property is not set in the connectionInfo